src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/InteropServices/ComWrappers.CoreRT.cs

```diff
@@ -0,0 +1,602 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Threading;
+using System.Runtime.CompilerServices;
+using System.Runtime.Versioning;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices
+{
+    /// <summary>
+    /// Internal enumeration used by the runtime to indicate the scenario for which ComWrappers is being used.
+    /// </summary>
+    internal enum ComWrappersScenario
+    {
+        Instance = 0,
+        TrackerSupportGlobalInstance = 1,
+        MarshallingGlobalInstance = 2,
+    }
+
+    /// <summary>
+    /// Class for managing wrappers of COM IUnknown types.
+    /// </summary>
+    public abstract partial class ComWrappers
+    {
+        const int DispatchAlignmentThisPtr = 16;
+        private static readonly ConditionalWeakTable<object, ManagedObjectWrapperHolder> CCWTable = new ConditionalWeakTable<object, ManagedObjectWrapperHolder>();
+
+        /// <summary>
+        /// ABI for function dispatch of a COM interface.
+        /// </summary>
+        public partial struct ComInterfaceDispatch
+        {
+            /// <summary>
+            /// Given a <see cref="System.IntPtr"/> from a generated Vtable, convert to the target type.
+            /// </summary>
+            /// <typeparam name="T">Desired type.</typeparam>
+            /// <param name="dispatchPtr">Pointer supplied to Vtable function entry.</param>
+            /// <returns>Instance of type associated with dispatched function call.</returns>
+            public static unsafe T GetInstance<T>(ComInterfaceDispatch* dispatchPtr) where T : class
+            {
+                // See the dispatch section in the runtime for details on the masking below.
+                const long DispatchThisPtrMask = ~0xfL;
```

Comment:
Do you mean that I should use `~(DispatchAlignmentThisPtr - 1)` because `DispatchAlignmentThisPtr` equals to 16, and I do not see a way how it can be done without calculations or declaring existing constant as it is.

Reply:
Yes. You can also drop the alignment tricks alltogether as I have commented in https://github.com/dotnet/runtimelab/pull/653#discussion_r578670169